src/java.base/share/classes/java/util/IdentityHashMap.java

```diff
@@ -267,9 +267,25 @@ private void init(int initCapacity) {
      * @throws NullPointerException if the specified map is null
      */
     public IdentityHashMap(Map<? extends K, ? extends V> m) {
-        // Allow for a bit of growth
-        this((int) ((1 + m.size()) * 1.1));
-        putAll(m);
+        this(m, m.size());
+    }
+
+    /**
+     * Constructs a new identity hash map containing the keys-value mappings
+     * in the specified map, with expected size
+     *
+     * @param map the map whose mappings are to be placed into this map
+     * @param expectedSize the expected size of the created IdentityHashMap
+     * @throws NullPointerException if the specified map is null
+     */
+    private IdentityHashMap(Map<? extends K, ? extends V> map, int expectedSize) {
```

Comment:
Why are you writing a new constructor when you can just change the old call to `this(m.size());`?

Reply:
because I don't like to call m.size() twice.
the original codes did so: one call at the `public IdentityHashMap(Map<? extends K, ? extends V> map)` , and the second inside of `putAll`.
well in most Map implementations `size()` seems O1, and returns a single int number field, but it actually defers in some Map implementations.

Reply:
and adding a private constructor should not cause any trouble I guess?

Reply:
> @liach implementations `size()` seems O1, and returns a single int number field, but it actually defers in some Map implementations. 

for example, in ConcurrentSkipListMap and ConcurrentHashMap, `size()` function is far complicated than reading a field, thus calling it twice meaninglessly is not a wise choice.

Reply:
Imo you should just remove the `if (expectedSize == 0)` check than using this somewhat ugly trick to avoid calling `size()` twice (the second call is only used for this relatively useless fast-path, especially for the concurrent collections you refer to)

Reply:
In fact, if we do worry about the performance of adding from maps, calling `map.forEach(this::put);` would be a better alternative both in concurrency (as the concurrent map itself takes charage) and object allocation-wise (no allocation of immutable entry objects), but that belongs to another issue.

Reply:
@liach Hi. please have a look at the latest commit.
do you think it be better now?

Reply:
Oops, didn't notice there was this helpful `init` method. Does look much more straightforward now.

Reply:
I'm starting to look at this again. First, a quick note -- I don't think there should be any IdentityHashMap changes here. That uses a completely different internal structure and allocation policy, and it's kind of a distraction from the main point of this change. (As the preceding chain of comments illustrates.)

There is a question of whether IdentityHashMap ought to "allow for some growth" when copying a map, but it's hard to see how much of a problem this really is. I'd suggest focusing on getting this change done, and then (if you still have energy) to try to eradicate the `(int) (expected / 0.75) + 1` idiom that's sprinkled around the JDK.

Reply:
> I'm starting to look at this again. First, a quick note -- I don't think there should be any IdentityHashMap changes here. That uses a completely different internal structure and allocation policy, and it's kind of a distraction from the main point of this change. (As the preceding chain of comments illustrates.)
> 
> There is a question of whether IdentityHashMap ought to "allow for some growth" when copying a map, but it's hard to see how much of a problem this really is. I'd suggest focusing on getting this change done, and then (if you still have energy) to try to eradicate the `(int) (expected / 0.75) + 1` idiom that's sprinkled around the JDK.

OK, will split this part from this pr.